locust/test/test_main.py

```diff
@@ -1766,34 +2128,58 @@ def t(self):
                     "-",
                     "--worker",
                 ],
-                stderr=STDOUT,
+                stderr=PIPE,
                 stdout=PIPE,
                 text=True,
+                bufsize=1,
             )
-            gevent.sleep(2)
             proc_worker2 = subprocess.Popen(
                 [
                     "locust",
                     "-f",
                     "-",
                     "--worker",
                 ],
-                stderr=STDOUT,
+                stderr=PIPE,
                 stdout=PIPE,
                 text=True,
+                bufsize=1,
             )
-            stdout = proc.communicate()[0]
-            stdout_worker = proc_worker.communicate()[0]
-            stdout_worker2 = proc_worker2.communicate()[0]
 
-            self.assertIn('All users spawned: {"User1": 1} (1 total users)', stdout)
-            self.assertIn("Shutting down (exit code 0)", stdout)
-            self.assertNotIn("Traceback", stdout)
-            self.assertNotIn("Traceback", stdout_worker)
-            self.assertNotIn("Traceback", stdout_worker2)
+            output = []
+
+            poll_until(lambda: all_users_spawned(proc, output), timeout=30)
+
+            stdout, stderr = proc.communicate(timeout=30)
+            stdout_worker, stderr_worker = proc_worker.communicate(timeout=10)
+            stdout_worker2, stderr_worker2 = proc_worker2.communicate(timeout=10)
+
+            output.extend(stdout.splitlines())
+            output.extend(stderr.splitlines())
+            output.extend(stdout_worker.splitlines())
+            output.extend(stderr_worker.splitlines())
+            output.extend(stdout_worker2.splitlines())
+            output.extend(stderr_worker2.splitlines())
+
+            self.assertTrue(
+                any("Shutting down" in line for line in output),
+                f"'Shutting down' not found in output:\n{''.join(output)}",
+            )
+            self.assertIn('All users spawned: {"User1": 1} (1 total users)', "\n".join(output))
+            self.assertNotIn("Traceback", "\n".join(output))
 
             self.assertEqual(0, proc.returncode)
             self.assertEqual(0, proc_worker.returncode)
+            self.assertEqual(0, proc_worker2.returncode)
+
+            for p in [proc, proc_worker, proc_worker2]:
```

Comment:
Why is this necessary? We have already checked that their return code was zero, so the processes must already be dead :)